### PR TITLE
remove DEBUG_BUILD_OPTIMIZATION_ONE

### LINKED ISSUE #283: Bad assembly with SimpleArray caster
In the [SimpleArrayCaster.hpp](https://github.com/solvcon/modmesh/pull/266) file, a custom caster for **SimpleArray** was defined to cast from **SimpleArrayPlex** to **SimpleArray**. By default, the debug build uses the **-O0** flag and the binary is incorrect. The compiler ignores the custom caster and opts for the generic caster instead. Setting the optimization level to **-O1** or higher leads to a correct binary and the custom caster works correctly.

This anomaly was identified on Windows (tested with MSVC 2022) and macOS (tested with LLVM 15 and GCC 13 toolchains). Notably, Ubuntu 22.04 with GCC 11 did not exhibit this issue.

The issue was initially observed in [PR#266](https://github.com/solvcon/modmesh/pull/266), and a concise investigation report is available [here](https://github.com/solvcon/modmesh/pull/266#issuecomment-1946362573).

As a temporary solution, a workaround involves adding a flag, **DEBUG_BUILD_OPTIMIZATION_ONE**, in CMake to ensure successful CI passes on Windows and macOS by applying **-O1** for the debug build. By default, this flag should be set to **OFF** to reproduce the issue. Developers are advised to apply this CMake flag to ensure correct behavior in their local environments, aligning with CI expectations.

### PR DESCRIPTION
**Pull request by tigercosmos**
To address #283, two steps should be applied.
- put the `SimpleArrayCaster.hpp` into `array_common.hpp` (check more at the comments in the code)
- refactored `wrap_modifiers()`, which leads some type issues for pybind11 when we introduce customized caster for the array.

close #283

### CODE REVIEW

File: .github/workflows/devbuild.yml [hunk 1]
```diff
@@ -477,26 +477,24 @@ jobs:
...
             -Dpybind11_DIR="$(pybind11-config --cmakedir)" `
-            -DDEBUG_BUILD_OPTIMIZATION_ONE=ON `
             -S${{ github.workspace }} `
             -B${{ github.workspace }}/build
           cmake --build ${{ github.workspace }}/build `
             --config ${{ matrix.cmake_build_type }} `
             --target ALL_BUILD
 
-      - name: cmake ALL_BUILD USE_PYTEST_HELPER_BINDING=ON DEBUG_BUILD_OPTIMIZATION_ONE=ON
+      - name: cmake ALL_BUILD USE_PYTEST_HELPER_BINDING=ON
```
Comment (tigercosmos) [inline .github/workflows/devbuild.yml:491]: removed all `DEBUG_BUILD_OPTIMIZATION_ONE` in the CI

File: CMakeLists.txt [hunk 2]
```diff
@@ -97,20 +92,8 @@ message(STATUS "DEBUG_SYMBOL: ${DEBUG_SYMBOL}")
 if(DEBUG_SYMBOL)
     if(MSVC)
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /DEBUG")
-        if(DEBUG_BUILD_OPTIMIZATION_ONE)
-            # MSVC optimization flags are different from GCC/Clang, so there is no -O1 in MSVC.
```
Comment (tigercosmos) [inline CMakeLists.txt:101]: Deleting these is a bit regrettable

File: cpp/modmesh/buffer/pymod/array_common.hpp [hunk 3]
```diff
@@ -33,6 +33,14 @@
 #include <modmesh/buffer/SimpleArray.hpp>
 #include <modmesh/buffer/pymod/TypeBroadcast.hpp>
 
+// We faced an issue where the template specialization for the caster of
+// SimpleArray<T> doesn't function correctly on both macOS and Windows.
+// While the root cause of the problem remains unclear, a workaround is
+// available by including the caster header in this file, impacting
+// wrap_SimpleArray.cpp.
+// See more details in the issue: https://github.com/solvcon/modmesh/issues/283
+#include <modmesh/buffer/pymod/SimpleArrayCaster.hpp>
```
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/array_common.hpp:42]: please see the comments.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp [hunk 4]
```diff
@@ -141,24 +141,15 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
             .def_property_readonly("nbody", &wrapped_type::nbody)
             .def_property_readonly("plex", [](wrapped_type const & arr)
                                    { return pybind11::cast(SimpleArrayPlex(arr)); })
-            .wrap_modifiers()
+            .def(
+                "fill", [](wrapped_type & arr, value_type const value)
```
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: it's interesting that we need to write in this way to make pybind11 work correctly.

File: cpp/modmesh/python/common.hpp [hunk 5]
```diff
@@ -285,15 +285,15 @@ WrapBase
     wrapper_type & METHOD(Args &&... args)                                    \
     {                                                                         \
         m_cls.METHOD(std::forward<Args>(args)...);                            \
-        return *static_cast<wrapper_type *>(this);                            \
+        return *static_cast<std::add_pointer_t<wrapper_type>>(this);          \
```
Comment (tigercosmos) [inline cpp/modmesh/python/common.hpp]: use modern C++

**Comment by tigercosmos** [comment]
@yungyuc ready for review.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp (hunk 4 above)
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: This is interesting, and I used some similar workarounds before with boost.python.

Does using lambda resolve the bad assembly?  If yes, please add comments here to explain to future maintainers.  Also, try to keep `wrap_modifiers()`.

File: cpp/modmesh/python/common.hpp (hunk 5 above)
Comment (yungyuc) [reply cpp/modmesh/python/common.hpp]: If this does nothing to the resolution of the bad assembly, please avoid the refactoring change in this **changeset**.

You may use another refactoring PR.

**Review by yungyuc — COMMENTED** [review]
Thanks for the update.

- [x] Discuss lambda.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp (hunk 4 above)
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: sure, kept `wrap_modifiers` and used lambda inside.

File: cpp/modmesh/python/common.hpp (hunk 5 above)
Comment (tigercosmos) [reply cpp/modmesh/python/common.hpp]: sure

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp (hunk 4 above)
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: 1. The lambda isn't responsible for the problematic assembly; instead, putting `SimpleArrayCaster.hpp` into `array_common.hpp` addresses it (I don't know why this works, to be honested)
2. Upon integrating `SimpleArrayCaster.hpp` into wrap_`SimpleArray.cpp`, a lambda becomes necessary to rectify the erroneous implicit casting in Pybind11. Without the lambda, Pybind11 fails to build due to certain type issues, albeit not explored in depth.

**Comment by tigercosmos** [comment]
@yungyuc  fixed the issues above.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp [hunk 6]
```diff
@@ -152,7 +152,12 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
         namespace py = pybind11; // NOLINT(misc-unused-alias-decls)
 
         (*this)
-            .def("fill", &wrapped_type::fill, py::arg("value"))
+            .def(
+                // Use lambda to avoid bad implicit conversion of SimpleArray<T> in pybind11
```
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: Please add the issue #283 in the comment for easy reference.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp [hunk 7]
```diff
@@ -152,7 +152,13 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
         namespace py = pybind11; // NOLINT(misc-unused-alias-decls)
 
         (*this)
-            .def("fill", &wrapped_type::fill, py::arg("value"))
+            .def(
+                // Use lambda to avoid bad implicit conversion of SimpleArray<T>in pybind11
+                // (see https://github.com/solvcon/modmesh/issues/283)
```
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp:157]: I tweaked the comment with the link to the issue #283 added.

**Review by yungyuc — APPROVED** [review]
LGTM.  Waiting for CI and I will merge.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp (hunk 4 above)
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: Oh this is even more interesting.  There is definitely something we missed or misunderstood in the language and/or the compiler.  But pragmatism beats purity.  Let's get this to work and keep the experience.  Maybe someday the Lord will shed lights on this matter.

This is good finding and description, and will be a good reference in the future.

**Comment by tigercosmos** [comment]
thanks @yungyuc

**Pull request merged**